{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The croissant was created with reference to the [mlcroissant-recipes-introduction-notebook](https://github.com/mlcommons/croissant/blob/main/python/mlcroissant/recipes/introduction.ipynb).\n",
    "\n",
    "Install the following packages if necessary (`mlcroissant` requires `python` 3.10+)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ! pip install mlcroissant charset-normalizer==3.1.0 gitpython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Locate the path first. Here we use `FileObjects` only."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mlcroissant as mlc\n",
    "\n",
    "# FileObjects and FileSets define the resources of the dataset.\n",
    "distribution = [\n",
    "    # Hi-TPH is hosted on a GitHub repository:\n",
    "    mlc.FileObject(\n",
    "        id=\"github-repository\",\n",
    "        name=\"github-repository\",\n",
    "        description=\"Hi-TPH repository on GitHub.\",\n",
    "        content_url=\"https://github.com/Jiadong001/Hi-TpH\",\n",
    "        encoding_format=\"git+https\",\n",
    "        sha256=\"main\",\n",
    "    ),\n",
    "    # Within that repository, level I-IV files are csv files with different kinds of columns:\n",
    "    mlc.FileObject(\n",
    "        id=\"level-I.csv\",                           # drop preffix in content_url\n",
    "        name=\"level-I.csv\",\n",
    "        description=\"Hi-TPH level I csv file is hosted on the GitHub repository.\",\n",
    "        content_url=\"data/Hi-TpH-level-I.csv\",      # path = contained_in + content_url\n",
    "        contained_in=[\"github-repository\"],\n",
    "        encoding_format=\"text/csv\",\n",
    "    ),\n",
    "    mlc.FileObject(\n",
    "        id=\"level-II.csv\",\n",
    "        name=\"level-II.csv\",\n",
    "        description=\"Hi-TPH level II A csv file is hosted on the GitHub repository.\",\n",
    "        content_url=\"data/Hi-TpH-level-II.csv\",\n",
    "        contained_in=[\"github-repository\"],\n",
    "        encoding_format=\"text/csv\",\n",
    "    ),\n",
    "    mlc.FileObject(\n",
    "        id=\"level-III.csv\",\n",
    "        name=\"level-III.csv\",\n",
    "        description=\"Hi-TPH level III csv file is hosted on the GitHub repository.\",\n",
    "        content_url=\"data/Hi-TpH-level-III.csv\",\n",
    "        contained_in=[\"github-repository\"],\n",
    "        encoding_format=\"text/csv\",\n",
    "    ),\n",
    "    mlc.FileObject(\n",
    "        id=\"level-IV.csv\",\n",
    "        name=\"level-IV.csv\",\n",
    "        description=\"Hi-TPH level IV csv file is hosted on the GitHub repository.\",\n",
    "        content_url=\"data/Hi-TpH-level-IV.csv\",\n",
    "        contained_in=[\"github-repository\"],\n",
    "        encoding_format=\"text/csv\",\n",
    "    )\n",
    "]\n",
    "distribution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`RecordSet`: Load records from one/many csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "components2desc_dict = {\n",
    "    \"antigen.epitope\": \"Peptide sequence of the antigen epitope.\",\n",
    "\n",
    "    \"alpha.cdr3\": \"Complementarity Determining Region 3 (CDR3) amino acid sequence of TCR alpha chain.\",\n",
    "    \"alpha.v\": \"V gene segment of TCR alpha chain.\",\n",
    "    \"alpha.j\": \"J gene segment of TCR alpha chain.\",\n",
    "    \"alpha.vseq.reconstructed\": \"Reconstructed variable domain sequence of TCR alpha chain using V/J gene annotations and CDR3 sequences.\",\n",
    "\n",
    "    \"beta.cdr3\": \"Complementarity Determining Region 3 (CDR3) amino acid sequence of TCR beta chain.\",\n",
    "    \"beta.v\": \"V gene segment of TCR beta chain.\",\n",
    "    \"beta.j\": \"J gene segment of TCR beta chain.\",\n",
    "    \"beta.vseq.reconstructed\": \"Reconstructed variable domain sequence of TCR beta chain using V/(D/)J gene annotations and CDR3 sequences.\",\n",
    "\n",
    "    \"hla.allele\": \"HLA allele name.\",\n",
    "    \"hla.full.seq\": \"Full HLA protein sequence.\",\n",
    "    \"hla.clip.seq\": \"Amino acid sequence of HLA alpha-1 and alpha-2 domains, which is clipped from the full HLA protein sequence.\",\n",
    "    \"hla.short.seq\": \"HLA pseudo sequence, extracted from HLA alpha-1 and alpha-2 domains.\"\n",
    "}\n",
    "\n",
    "level2components_dict = {\n",
    "    \"level-I\": [\"antigen.epitope\", \"beta.cdr3\"],\n",
    "\n",
    "    \"level-II\": [\"antigen.epitope\", \"hla.allele\", \"beta.cdr3\",\n",
    "                \"hla.full.seq\", \"hla.clip.seq\", \"hla.short.seq\"],\n",
    "\n",
    "    \"level-III\": [\"antigen.epitope\", \"hla.allele\", \"alpha.cdr3\", \"beta.cdr3\",\n",
    "                \"hla.full.seq\", \"hla.clip.seq\", \"hla.short.seq\"],\n",
    "\n",
    "    \"level-IV\": [\"antigen.epitope\", \"hla.allele\", \n",
    "                \"alpha.v\", \"alpha.j\",\"alpha.cdr3\", \n",
    "                \"beta.v\", \"beta.j\", \"beta.cdr3\",\n",
    "                \"alpha.vseq.reconstructed\", \"beta.vseq.reconstructed\",\n",
    "                \"hla.full.seq\", \"hla.clip.seq\", \"hla.short.seq\"],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "record_sets = []\n",
    "for level, components in level2components_dict.items():\n",
    "    record_set = mlc.RecordSet(\n",
    "            id=level,\n",
    "            name=level,\n",
    "            # Each record has one or many fields...\n",
    "            fields=[\n",
    "                # Fields can be extracted from the FileObjects/FileSets.\n",
    "                mlc.Field(\n",
    "                    id=f\"{level}/{component}\",              # e.g. \"level-I/antigen.epitope\"\n",
    "                    name=component,\n",
    "                    description=components2desc_dict[component],\n",
    "                    data_types=mlc.DataType.TEXT,\n",
    "                    source=mlc.Source(\n",
    "                        file_object=f\"{level}.csv\",\n",
    "                        # Extract the field from the column of a FileObject/FileSet:\n",
    "                        extract=mlc.Extract(column=component),      \n",
    "                    ),\n",
    "                ) \n",
    "                for component in components\n",
    "            ],\n",
    "        )\n",
    "    print(record_set.fields)\n",
    "    record_sets.append(record_set)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = mlc.Metadata(\n",
    "    name=\"Hi-TPH\",\n",
    "    # Descriptions can contain plain text or markdown.\n",
    "    description=(\n",
    "        \"A Large-Scale Hierarchical Dataset for TCR-pHLA Binding Prediction.\"\n",
    "    ),\n",
    "    cite_as=None,                   # None: to be updated\n",
    "    license=\"https://creativecommons.org/licenses/by-nc/4.0/\",\n",
    "    version=None,\n",
    "    date_published=None,\n",
    "    url=\"https://github.com/Jiadong001/Hi-TpH\",\n",
    "    distribution=distribution,\n",
    "    record_sets=record_sets,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(metadata.issues.report())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save croissant (metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "with open(\"croissant.json\", \"w\") as f:\n",
    "    content = metadata.to_json()\n",
    "    content = json.dumps(content, indent=2)\n",
    "    print(content)\n",
    "    f.write(content)\n",
    "    f.write(\"\\n\")       # Terminate file with newline"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py310",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.1.undefined"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
